sparta/sparta/resources/PriorityQueue.hpp

```diff
@@ -0,0 +1,186 @@
+         *        sorting algorithm
+         */
+        PriorityQueue() :
+            priority_items_(bounded_cnt)
+        {}
+
+        /**
+         * \brief Create a priority queue with a specific instance of the
+         *        sorting algorithm
+         * \param sort_alg Reference to the sorting algorithm instance
+         */
+        PriorityQueue(const SortingAlgorithmT & sort_alg) :
+            priority_items_(bounded_cnt),
+            sort_alg_(sort_alg)
+        {}
+
+        /**
+         * \brief Inserts the data item into the list using the
+         *        sorting alg.  Stops at the first insertion.
+         * \param data The data to insert
+         */
+        void insert(const DataT & data)
+        {
+            const auto eit = priority_items_.end();
+            for(auto it = priority_items_.begin(); it != eit; ++it)
+            {
+                if(sort_alg_(data, *it)) {
+                    priority_items_.insert(it, data);
+                    return;
+                }
+            }
+            priority_items_.emplace_back(data);
+        }
+
+        //! Get the number of items in the queue
+        size_t size() const {
+            return priority_items_.size();
+        }
+
+        //! Is the queue empty?
+        bool empty() const {
+            return priority_items_.empty();
+        }
+
+        //! Get the first element in the queue
+        const DataT & top() const {
+            sparta_assert(false == empty(), "Grabbing top from an empty queue");
+            return priority_items_.front();
+        }
+
+        //! Get the last element (lowest priority) in the queue
+        const DataT & back() const {
+            sparta_assert(false == empty(), "Grabbing back from an empty queue");
+            return priority_items_.back();
+        }
+
+        //! Pop the front of the queue (highest priority)
+        void pop() {
+            sparta_assert(false == empty(), "Popping on an empty priority queue");
+            priority_items_.pop_front();
+        }
+
+        //! Clear the entire queue
+        void clear() {
+            priority_items_.clear();
+        }
+
+        //! Remove the item from the queue
+        void remove(const DataT & data) {
+            priority_items_.remove(data);
+        }
+
+        //! Erase the item from the queue (const_iterator/iterator)
+        void erase(const const_iterator & it) {
+            priority_items_.erase(it);
+        }
+
+        /**
+         * \brief Force a data entry to the front of the queue
+         * \param data Reference to the data object forced to the
+         *             front of the queue
+         *
+         * Push the data item to the front of the queue, bypassing the
+         * internal sorting algorithm.  This will mess up the sorting
+         * -- be warned.
+         */
+        void forceFront(const DataT & data) {
+            priority_items_.emplace_front(data);
+        }
```

Comment:
Interesting. Wondering the usage? If an entry can be pushed like this way, a `refresh` method to sort up entries might sound useful for users.

Reply:
I thought about a `resort` call, but I'd rather someone request it.  :smile: 

This method is useful in situations like coherency models, where multiple requests from many sources might come in and be sorted properly.  But then, at the last minute, an I/O device really needs attention due to a possible buffer overflow (I'm making stuff up :wink: ) and must be highest priority even though it actually can be considered lower.